db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/fk/GraphAnnotator.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2012 - 2020 Splice Machine, Inc.
+ * This file is part of Splice Machine.
+ * Splice Machine is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3, or (at your option) any later version.
+ * Splice Machine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.splicemachine.db.iapi.sql.dictionary.fk;
+
+import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.db.iapi.reference.SQLState;
+import com.splicemachine.utils.Pair;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class GraphAnnotator {
+    private final Graph graph;
+    private final String newConstraintName;
+
+    class Path {
+        public Path(List<Integer> vertices, EdgeNode.Type action) {
+            this.vertices = vertices;
+            this.action = action;
+        }
+
+        public String toString(Graph g) {
+            StringBuilder sb = new StringBuilder();
+            sb.append("PATH action: ").append(action).append(" ");
+            for(int v : vertices) {
+                sb.append(g.getName(v)).append(" ");
+            }
+            return sb.toString();
+        }
+
+        List<Integer> vertices;
+        EdgeNode.Type action;
+
+        private List<Integer> removeSurrogates(List<Integer> in) {
+            return in.stream().map((vertex) -> {
+                if (graph.isSurrogate(vertex)) {
+                    return graph.getOriginal(vertex);
+                } else {
+                    return vertex;
+                }
+            }).collect(Collectors.toList());
+        }
+
+        public boolean intersects(Path needle) {
+            List<Integer> originalPathVertices = removeSurrogates(vertices);
+            List<Integer> originalNeedleVertices = removeSurrogates(needle.vertices);
+
+
+            return originalPathVertices.stream()
+                    .distinct()
+                    .filter(originalNeedleVertices::contains)
+                    .collect(Collectors.toSet()).size() > 0;
+        }
+    }
+
+    static class Annotation {
+        List<Path> paths;
+        public Annotation() {
+            paths = new ArrayList<>();
+        }
+    }
+
+    Annotation[] annotations;
+
+    public GraphAnnotator(String newConstraintName, Graph graph) {
```

Comment:
this also looks easily unit-testable

Reply:
True, planning for extensive testing of the graph and its algorithms in [this JIRA](https://splicemachine.atlassian.net/browse/DB-10719), that is actually why I did the separation of the graph checker from the rest of the data dictionary so I can unit test it.